text/0000-stack-buildpacks.md

```diff
@@ -0,0 +1,671 @@
+* a Ruby app requires the `libpq` mixin and the `db-cert`
+* a Java app requires only the `db-cert`
+* a Python app requires the `db-cert` and the stack already provides the `libpq` mixin
+
+The example stackpack will be skipped after detection in the following cases:
+* an app requires the `libpq` mixin but not the `db-cert`
+* an app requires neither the `libpq` mixin nor the `db-cert`
+
+### Example: Chrome and Chromedriver
+
+As another example of these resolution rules, consider a stackpack that provides the `google-chrome-stable` mixin and a build plan dependency called `chromedriver`. During the detect phase, the stackpack will pass detection in the following cases:
+* an app requires the `google-chrome-stable` mixin and the `chromedriver`
+* an app requires only the `chromedriver`
+* an app requires the `chromedriver` and the stack already provides the `google-chrome-stable` mixin
+
+The example stackpack will be skipped after detection in the following cases:
+* an app requires the `google-chrome-stable` mixin but not the `chromedriver`
+* an app requires neither the `google-chrome-stable` mixin nor the `chromedriver`
+
+During the build phase the example stackpack will be given a [buildpack plan](https://github.com/buildpacks/spec/blob/main/buildpack.md#buildpack-plan-toml) containing the entries it must provide. For example, if both dependencies are required it will receive:
+
+```toml
+[[entries]]
+mixin = "google-chrome-stable"
+
+[[entries]]
+name = "chromedriver"
+```
+
+If only the `chromedriver` is required, or the `google-chrome-stable` is already provided by the stack, the build phase would receive:
+
+```toml
+[[entries]]
+name = "chromedriver"
+```
+
+If the stack does not already provide `google-chrome-stable`, the stackpack may install it anyways. This prevents the case where an app may not work at runtime if that app requires only the `chromedriver` and the `google-chrome-stable` is not provided by the stack.
+
+## Caching and Restoring
+
+During the export phase, the lifecycle will store any excluded cached stackpack layers created during the build or extend phase in the cache.
+
+During the restore phase of the next build, the lifecycle will download the excluded cached stackpack layers with the cache and store them as tarballs in the `<layers>` directory and the `<run-layers>` directory (i.e. they will not be extracted). The `restorer` cannot extract the these layers because it will not run with root privileges. In addition, the `restorer` may run in a different container than the build, which means changes made to the base image are not guaranteed to carry forward. The `builder` will run with root privileges and untar these layers and cache, but will drop privileges after executing stackpacks and before running application buildpacks. The `extender` will also run with root privileges and untar the cache.
```

Comment:
```suggestion
During the restore phase of the next build, the lifecycle will download the excluded cached stackpack layers with the cache and store them as tarballs in the `<layers>` directory and the `<run-layers>` directory (i.e. they will not be extracted). The `restorer` cannot extract these layers because it will not run with root privileges. In addition, the `restorer` may run in a different container than the build, which means changes made to the base image are not guaranteed to carry forward. The `builder` will run with root privileges and untar these layers and cache, but will drop privileges after executing stackpacks and before running application buildpacks. The `extender` will also run with root privileges and untar the cache.
```